### ISSUE TITLE
ClassNaming rule treating Kotlin syntax ` as part of class name

### ISSUE THREAD

**Issue by hqzxzwb** [detekt/detekt#2656]
## Expected Behavior
ClassNaming rule should only check class name and should not check syntax \`

## Observed Behavior
ClassNaming rule treating Kotlin syntax \` as part of class name

## Steps to Reproduce
```Kotlin
class `$Test$`
```
The above code is warned about a ClassNaming issue with default config 
```
  ClassNaming:
    active: true
    excludes: ['**/test/**', '**/androidTest/**', '**/*.Test.kt', '**/*.Spec.kt', '**/*.Spek.kt']
    classPattern: '[A-Z$][a-zA-Z0-9$]*'
```
When I change the classPattern to ``[`A-Z$][`a-zA-Z0-9$]*``, the warning is gone.
This means detekt is treating the syntax ` as part of class name, which is probably not intended.

## Context
Class names with $ in it should pass the test, otherwise you would not write it to the default pattern. But it's not possible to have such a class name without the \` syntax.

## Your Environment

* Version of detekt used: 1.8.0
* Version of Gradle used (if applicable): 5.6.4
* Operating System and version: macOS Catalina
* Link to your project (if it's a public repository):

**Comment by arturbosch** [comment]
We could ignore names with this syntax altogether or implement this as an option.
Using this syntax is quite intentional, isn't it ?

**Comment by hqzxzwb** [comment]
Yes it is quite intentional. Personally I use it to create special class names when I do not want anyone to use the class but it has to be public for some other reason.

**Comment by BraisGabin** [comment]
I just realised. Why do we allow classes with `$` in the name? It has little sense. The `$` is usually used by the annotation processors to create new classes. And they use that symbol to get a name that is not going to conflict with the user class names. Our tests didn't check the use of `$` neighter. And you can't use te symbol `$` if you don't use the char `. I think that it's safe to remove it.

**Comment by hqzxzwb** [comment]
So remove $ from default pattern and add an option to ignore class names in \`?

**Comment by BraisGabin** [comment]
I created a PR removing the `$` because right now you can't use it even with the current pattern. And I think that we don't need an option here. If you want to allow that case you can use this pattern: ``([A-Z][a-zA-Z0-9]*)|(`.*`)``

In my opinion the use of ` in the class names should not be "encoraged". If we add an option allowing that we are kind of econraging it. If you have a reason to use ` in your names you can do it with the previous pattern. What do you thing?

**Comment by hqzxzwb** [comment]
It seems a bit weird to see the \` syntax as part of the class name and match it in the pattern. I actually took quite a while to find this out. It is not a part of class name. But if the \` syntax is not treated as part of class name, we can't tell it apart from normal class names, that is why I suggested an option.

**Comment by hqzxzwb** [comment]
I am thinking of this if not encouraging free class naming and not treating \` as part of class name:
```
classPattern: '[A-Z][a-zA-Z0-9]*'
backQuotedClassPattern:''
```
Default empty pattern effectively forbids back quoted class names because class names cannot be empty. And users can manually change the pattern to meet their needs.

**Comment by BraisGabin** [comment]
We are using my proposal right now in the [FunctionNaming rule](https://detekt.github.io/detekt/naming.html#configuration-options-6). What do you think, @arturbosch?

**Comment by schalkms** [comment]
Closed by 924104e

**Comment by BraisGabin** [comment]
We are talking about adding (or not) the `backQuotedClassPattern` option. That PR is related but it's not the same.

**Comment by arturbosch** [comment]
> We are using my proposal right now in the [FunctionNaming rule](https://detekt.github.io/detekt/naming.html#configuration-options-6). What do you think, @arturbosch?

You mean `excludeClassPattern` right? Sounds good to me.
Do we want to have `\`.*\`` as a default pattern for these config properties?

**Comment by BraisGabin** [comment]
> You mean `excludeClassPattern` right? Sounds good to me.

Yes, I mean `ClassNaming>classPattern`.

> Do we want to have `` `.*` `` as a default pattern for these config properties?

I don't think so. If we add that there is like we are encoraging that type of naming for classes and I think that we should not. I think that we should follow the [Coding Conventions](https://kotlinlang.org/docs/reference/coding-conventions.html#naming-rules):

> Names of classes and objects start with an upper case letter and use the camel case:

I mean, we should support this kind of cases, as we do with `` `.*` ``. But I don't think that we should add it as a default value.
